### ISSUE TITLE
Fatal error with conv layers in Windows 10 with GPU

### ISSUE THREAD

**Issue by JeffHebert** [rstudio/keras#31]
I keep getting a fatal error when running convolutional layers. I am able to build models with dense, dropout, activation, and simple_rnn layers. I've included information on the R session, the model, and GPU. I've successfully installed CUDA 8.0 and cudnn 5.1 for CUDA 8.0.

PS. I am quite excited that you brought Keras and Tensorflow to R!!! Thanks!

sessionInfo()
> R version 3.4.0 (2017-04-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)
Matrix products: default
locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    
attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     
other attached packages:
[1] keras_0.3.6
loaded via a namespace (and not attached):
 [1] compiler_3.4.0   magrittr_1.5     R6_2.2.1         tools_3.4.0      withr_1.0.2      Rcpp_0.12.10    
 [7] reticulate_0.8   memoise_1.1.0    tensorflow_0.8.2 jsonlite_1.5     digest_0.6.12    devtools_1.12.0 

Model
> Layer (type)                                          Output Shape                                     Param #            
conv2d_1 (Conv2D)                                     (None, 26, 26, 32)                               320                
conv2d_2 (Conv2D)                                     (None, 24, 24, 64)                               18496              
max_pooling2d_1 (MaxPooling2D)                        (None, 12, 12, 64)                               0                  
dropout_1 (Dropout)                                   (None, 12, 12, 64)                               0                  
flatten_1 (Flatten)                                   (None, 9216)                                     0                  
dense_1 (Dense)                                       (None, 128)                                      1179776            
dropout_2 (Dropout)                                   (None, 128)                                      0                  
dense_2 (Dense)                                       (None, 10)                                       1290               
===
Total params: 1,199,882.0
Trainable params: 1,199,882.0
Non-trainable params: 0.0

Device 0: "GeForce GTX 970"
> CUDA Driver Version / Runtime Version          8.0 / 8.0
CUDA Capability Major/Minor version number:    5.2
Total amount of global memory:                 4096 MBytes (4294967296 bytes)
(13) Multiprocessors, (128) CUDA Cores/MP:     1664 CUDA Cores
GPU Max Clock rate:                            1178 MHz (1.18 GHz)
Memory Clock rate:                             3505 Mhz
Memory Bus Width:                              256-bit
L2 Cache Size:                                 1835008 bytes

**Comment by jjallaire** [comment]
Please include code that will allow me to reproduce (smallest/simplest code
that exhibits the error pls!)

**Comment by JeffHebert** [comment]
I've had this error with my own code, and then I started working with the example code [mnist_cnn.R](https://rstudio.github.io/keras/articles/examples/mnist_cnn.html).

Here is my slightly modified version of that code


library(keras)

batch_size <- 128
num_classes <- 10
epochs <- 12

img_rows <- 28
img_cols <- 28

mnist <- dataset_mnist()
x_train <- mnist$train$x
y_train <- mnist$train$y
x_test <- mnist$test$x
y_test <- mnist$test$y

input_shape <- c(img_rows, img_cols, 1)
x_train <- array(as.numeric(x_train), dim = c(dim(x_train)[[1]], input_shape))
x_test <- array(as.numeric(x_test), dim = c(dim(x_test)[[1]], input_shape))

x_train <- x_train / 255
x_test <- x_test / 255

cat('x_train_shape:', dim(x_train), '\n')
cat(dim(x_train)[[1]], 'train samples\n')
cat(dim(x_test)[[1]], 'test samples\n')

y_train <- to_categorical(y_train, num_classes)
y_test <- to_categorical(y_test, num_classes)

model <- keras_model_sequential()
model %>%
  layer_conv_2d(filters = 32, kernel_size = c(3,3), activation = 'relu',
                input_shape = input_shape) %>% 
  layer_conv_2d(filters = 64, kernel_size = c(3,3), activation = 'relu') %>% 
  layer_max_pooling_2d(pool_size = c(2, 2)) %>% 
  layer_dropout(rate = 0.25) %>% 
  layer_flatten() %>% 
  layer_dense(units = 128, activation = 'relu') %>% 
  layer_dropout(rate = 0.5) %>% 
  layer_dense(units = num_classes, activation = 'softmax')

model %>%
  compile(
    loss = loss_categorical_crossentropy,
    optimizer = optimizer_adadelta(),
    metrics = c('accuracy')
  )


model %>% fit(x_train, y_train,
              batch_size = batch_size,
              epochs = epochs,
              verbose = 1,
              validation_data = list(x_test, y_test)
)

**Comment by jjallaire** [comment]
Okay, thanks. I'll run this on my Windows GPU machine tomorrow and
hopefully be able to reproduce!

**Comment by jjallaire** [comment]
Unfortunately your example runs fine on my Windows GPU installation (GPU info below, I noticed that my CUDA Capability version was 6.1 whereas yours is 5.2, not sure if that has anything to do with it though). I am running the latest public release of TensorFlow (v1.1).

GPU installation/configuration is of course notoriously finicky so there could be any number of things wrong here. 

```
 CUDA Device Query (Runtime API) version (CUDART static linking)

Detected 1 CUDA Capable device(s)

Device 0: "GeForce GTX 1080"
  CUDA Driver Version / Runtime Version          8.0 / 8.0
  CUDA Capability Major/Minor version number:    6.1
  Total amount of global memory:                 8192 MBytes (8589934592 bytes)
  (20) Multiprocessors, (128) CUDA Cores/MP:     2560 CUDA Cores
  GPU Max Clock rate:                            1734 MHz (1.73 GHz)
  Memory Clock rate:                             5005 Mhz
  Memory Bus Width:                              256-bit
  L2 Cache Size:                                 2097152 bytes
  Maximum Texture Dimension Size (x,y,z)         1D=(131072), 2D=(131072, 65536), 3D=(16384, 16384, 16384)
  Maximum Layered 1D Texture Size, (num) layers  1D=(32768), 2048 layers
  Maximum Layered 2D Texture Size, (num) layers  2D=(32768, 32768), 2048 layers
  Total amount of constant memory:               65536 bytes
  Total amount of shared memory per block:       49152 bytes
  Total number of registers available per block: 65536
  Warp size:                                     32
  Maximum number of threads per multiprocessor:  2048
  Maximum number of threads per block:           1024
  Max dimension size of a thread block (x,y,z): (1024, 1024, 64)
  Max dimension size of a grid size    (x,y,z): (2147483647, 65535, 65535)
  Maximum memory pitch:                          2147483647 bytes
  Texture alignment:                             512 bytes
  Concurrent copy and kernel execution:          Yes with 2 copy engine(s)
  Run time limit on kernels:                     Yes
  Integrated GPU sharing Host Memory:            No
  Support host page-locked memory mapping:       Yes
  Alignment requirement for Surfaces:            Yes
  Device has ECC support:                        Disabled
  CUDA Device Driver Mode (TCC or WDDM):         WDDM (Windows Display Driver Model)
  Device supports Unified Addressing (UVA):      Yes
  Device PCI Domain ID / Bus ID / location ID:   0 / 1 / 0
  Compute Mode:
     < Default (multiple host threads can use ::cudaSetDevice() with device simultaneously) >

deviceQuery, CUDA Driver = CUDART, CUDA Driver Version = 8.0, CUDA Runtime Version = 8.0, NumDevs = 1, Device0 = GeForce GTX 1080
Result = PASS
```

**Comment by jjallaire** [comment]
You could also try executing the equivalent python script in the same TensorFlow environment. Here it is:

```python
'''Trains a simple convnet on the MNIST dataset.

Gets to 99.25% test accuracy after 12 epochs
(there is still a lot of margin for parameter tuning).
16 seconds per epoch on a GRID K520 GPU.
'''

from __future__ import print_function

import tensorflow.contrib.keras.api.keras as keras
from tensorflow.contrib.keras.api.keras.datasets import mnist
from tensorflow.contrib.keras.api.keras.models import Sequential
from tensorflow.contrib.keras.api.keras.layers import Dense, Dropout, Flatten
from tensorflow.contrib.keras.api.keras.layers import Conv2D, MaxPooling2D

batch_size = 128
num_classes = 10
epochs = 12

# input image dimensions
img_rows, img_cols = 28, 28

# the data, shuffled and split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()


x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

```

**Comment by jjallaire** [comment]
Interestingly, when I execute the above script from within R via:

```r
reticulate::py_run_script('mnist_cnn.py')
```

Then everything works fine. However if I go to the terminal (again in the same Python/TF environment) I get a crash with `could not create cudnn handle: CUDNN_STATUS_NOT_INITIALIZED`

This to me indicates that there is something finicky about getting CUDNN setup properly that can cause a crash.

In any case, I think the issue you are seeing is almost certainly a software/hardware configuration issue rather than something about the keras R package per-se (note that the package is an extremely thin layer over the Python API so it is fairly unlikely to introduce bugs or incompatibilities idiosyncratic to the R interface).

**Comment by JeffHebert** [comment]
Hi jjallaire. Thanks for working on this while I was occupied (working) this week. You were exactly right about the CUDNN setup. I remember that I had to install and reinstall a few times. I realized there might be an old dll in the path, and sure enough I found cudnn64_5.dll for CUDA v7.5. I simply renamed the file and all my KERAS models are running.

```
C:\Users\jeff->where cudnn64_5.dll
C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.5\bin\cudnn64_5.dll
C:\Apps\cuda-8.0\bin\cudnn64_5.dll

```

THANK YOU for bringing this to R!
